### ISSUE TITLE
Boot 64 bit Windows on 32 Bit UEFI?

### ISSUE THREAD

**Issue by archeYR** [maharmstone/quibble#2]
There are some tablets/cheap laptops with an x64 capable processor, that are crippled by 32 bit UEFI firmware, thus these can boot 32 bit Windows only. Could it be possible to load 64 bit Windows on 32 bit version of quibble (reverse of what is mentioned in To-do list)? Thank you for your work in any case.

**Comment by maharmstone** [comment]
I wasn't aware of that! Yes, if that's happening, it'd make sense to have amd64-on-x86 support as well.

**Comment by Manouchehri** [comment]
If you’re curious, I wrote about the backstory about half a decade ago: https://github.com/Manouchehri/vi8/issues/4#issuecomment-135216613

**Comment by amluto** [comment]
(I’m a Linux kernel person.)

Linux supports 64-bit kernels on 32-bit UEFI. There is special kernel support for runtime services, and I doubt Windows has this support. The best way to support this on Windows is probably to tell Windows that runtime services are not available.

**Comment by 1nikolas** [comment]
If you were wondering why these cheap laptops/tablets can boot only 32 bit uefi [here](https://software.intel.com/en-us/blogs/2015/07/22/why-cheap-systems-run-32-bit-uefi-on-x64-systems) is a good explanation from Intel themselves. Essentially Intel makes cheap motherboards for computer manufactures which are prebuilt with ram, soc (integrated cpu), soundcard etc. So what they say in the blog post is that a cheap laptop with 32gb/64gb emmc and 2gb of ram can't benefit from a 64bit os. This is kinda true but since most apps is now dropping 32bit support, there is actually a benefit of doing that. If we can manage to boot 64 bit Windows on a locked firmware which only supports 32 bit uefi, we can install a lighter custom version of Windows 10 or even Windows 7/8. But is it possible though?

**Comment by maharmstone** [comment]
It's definitely theoretically possible - there's no problem booting a 64-bit OS from a 16-bit BIOS, after all. It'll just take some work.

From the article:
> This means your Black Friday special purchase with 2GB of RAM and 16-32GB of internal flash storage doesn't see any 64-bit architecture benefits.

This is flat-out wrong. Off the top of my head, you have more registers, you have the lack of segmentation speeding things up slightly, you have RIP-relative addressing speeding up rebasing, and you have a faster calling convention. Apart from that...

**Comment by z4v4l** [comment]
https://docs.microsoft.com/en-us/windows-hardware/drivers/bringup/uefi-requirements-that-apply-to-all-windows-platforms

Windows expects Runtime Services. And apparently their bitness must be the same as the OS. (and it's a reasonable assumption. not counting cases, when it's made due "because we can" reasons. Windows also could have "da speshul kernal support" for mismatched bitness RTS (after all, there is WoW) but why could it be needed except to show off?)

**Comment by maharmstone** [comment]
Yep, I've already twigged that. If you were booting 64-bit Windows from 32-bit UEFI or vice versa, you'd have to forego Runtime Services and convince Windows it was doing an old-fashioned BIOS boot. That's the way Quibble worked through most of its development, so that's not an issue.

**Comment by ShadowEO** [comment]
It should also be noted that a lot of these machines also lack CSM support (such as in my case with the NuVision TMAX TM101W638L) and loads of vendors refuse to update their BIOS since Connected Standby now is supported on 64-bit. 

If you can pull this feat off, I'd be glad to test it out as I've been longing to run 64-bit Windows on this machine for WSL support (and my guilty pleasure, WoW, which dropped 32-bit support two years ago).

**Comment by dimasirfandanu** [comment]
is this released yet?

**Comment by sidliu2000** [comment]
I have an older onda v919 z3975 atom that is similar. It is still a very capable machine with 4 core. would be great to run 64 bit windows or linux or even chromeos. But UEFI is crippled to 32 bit.

**Comment by 1nikolas** [comment]
> I have an older onda v919 z3975 atom that is similar. It is still a very capable machine with 4 core. would be great to run 64 bit windows or linux or even chromeos. But UEFI is crippled to 32 bit.

Running 64bit Windows would depend on this project.

Linux you can run 64bit (in fact you can't run 32bit linux on uefi afaik). On most of the distros you will need to add an efi file to the installer and then boot again to the live cd after the install to add that file to your installation in order for it to boot (you dont need to do this on fedora official and arch custom builds). 

Chromeos you can run 32bit only but in my case, eventhough the laptop I have has a chromeos version, it doesn't boot...

Btw I have a Lenovo 100s-11iby

**Comment by sidliu2000** [comment]
Why chromeos can run only 32 bit? Chromeos should be 64 bit only by now.

**Comment by 1nikolas** [comment]
> 
> 
> Why chromeos can run only 32 bit? Chromeos should be 64 bit only by now.

I mean in a locked firmware where you can only boot 32bit uefi. There is no trick to boot the 64bit version afaik

**Comment by sidliu2000** [comment]
I am hoping this Quibble project might be able to help.

**Comment by 1nikolas** [comment]
> 
> 
> I am hoping this Quibble project might be able to help.

The readme says:
> Quibble is the custom Windows bootloader 

So it won't help :(

**Comment by phuclv90** [comment]
> (I’m a Linux kernel person.)
> 
> Linux supports 64-bit kernels on 32-bit UEFI. There is special kernel support for runtime services, and I doubt Windows has this support. The best way to support this on Windows is probably to tell Windows that runtime services are not available.

Indeed to run 64-bit Linux on a 32-bit UEFI a [thunking layer](https://kernelnewbies.org/Linux_3.15#EFI_64-bit_kernels_can_be_booted_from_32-bit_firmware) is needed to convert things between the 32 and 64-bit ABIs. The kernel needs to be [built with `CONFIG_EFI_MIXED=y`](https://unix.stackexchange.com/a/180966/44425) to have that capability. But since Windows has always require the bitness of the OS the match that of UEFI, I don't think MS will spend time creating that thunking layer for Windows

**Comment by binarymaster** [comment]
> I don't think MS will spend time creating that thunking layer for Windows

I see this problem bothers a lot of people, so I think this problem will be addressed in ReactOS at some point.

**Comment by phuclv90** [comment]
Anyway I don't know what this has to do with quibble, because it's a mainly kernel thing instead of bootloader. If it's just because of the bootloader I guess you can just boot up TianoCore then boot 64-bit Windows from it without problem. I'm not sure if it TianoCore currently supports booting from 32-bit UEFI or not but I've did that to boot 64-bit Windows from GPT disk on a BIOS system. There may be other methods as well: [Installing Windows x64 on 32-bit UEFI (EFI-IA32) via GRUB](https://superuser.com/q/1055305/241386)

**Comment by 1nikolas** [comment]
> 
> 
> Anyway I don't know what this has to do with quibble, because it's a mainly kernel thing instead of bootloader. If it's just because of the bootloader I guess you can just boot up TianoCore then boot 64-bit Windows from it without problem. I'm not sure if it TianoCore currently supports booting from 32-bit UEFI or not but I've did that to boot 64-bit Windows from GPT disk on a BIOS system. There may be other methods as well: [Installing Windows x64 on 32-bit UEFI (EFI-IA32) via GRUB](https://superuser.com/q/1055305/241386)

that method doesn't work unfortunately. It just tells you to chainload the windows x64 boot file from grub which throws an error

**Comment by ShadowEO** [comment]
> Anyway I don't know what this has to do with quibble, because it's a mainly kernel thing instead of bootloader. If it's just because of the bootloader I guess you can just boot up TianoCore then boot 64-bit Windows from it without problem. I'm not sure if it TianoCore currently supports booting from 32-bit UEFI or not but I've did that to boot 64-bit Windows from GPT disk on a BIOS system. There may be other methods as well: [Installing Windows x64 on 32-bit UEFI (EFI-IA32) via GRUB](https://superuser.com/q/1055305/241386)

This method doesn't work, it **only** works with Grub, I've tried it multiple different ways and can't get Grub EFI to chainload a 64-bit Windows EFI executable. Your suggestion for TianoCore also doesn't help those that can't boot into CSM mode, such as these tablets that are locked into IA32 UEFI only (despite having a 64-bit capable Atom in them). Without CSM mode, you can't boot into Duet which is a requirement for that method.

**Comment by Prototyped** [comment]
Lenovo's ThinkPad 8 from 2014, then reviewed as [the Rolls Royce of small Windows tablets](https://betanews.com/2014/04/06/lenovo-thinkpad-8-the-rolls-royce-of-small-windows-tablets-review/), has several SKUs that are stuck with a 32-bit-only, no-CSM UEFI. These Bay Trail tablets support Connected Standby, presumably why they launched with 32-bit only UEFI. They only have 2 GB RAM, so the utility of 64-bit Windows (which is considerably more memory hungry than 32-bit Windows) is limited, but it would be cool to at least have the option.

I have booted the Lubuntu 64-bit live ISO off the 32-bit UEFI, so Linux is clearly fine with it. Windows definitely isn't.

**Comment by ShadowEO** [comment]
@Prototyped Linux has the ability to take over Runtime Services within it's kernel, essentially replacing the running UEFI implementation with itself. Grub sets this up and then the kernel finishes it. Since the kernel essentially takes the UEFI environment out of play, it can boot itself in 64-bit. Windows however doesn't have this ability.

**Comment by TheMadHau5** [comment]
> 
> 
> @Prototyped Linux has the ability to take over Runtime Services within it's kernel, essentially replacing the running UEFI implementation with itself. Grub sets this up and then the kernel finishes it. Since the kernel essentially takes the UEFI environment out of play, it can boot itself in 64-bit. Windows however doesn't have this ability.

May I know how and where this is implemented?

**Comment by phuclv90** [comment]
> @Prototyped 
> May I know how and where this is implemented?

It's a thunking layer that MS simply just doesn't bother spending effort writing. [Here are the related commits](https://kernelnewbies.org/Linux_3.15#EFI_64-bit_kernels_can_be_booted_from_32-bit_firmware)

**Comment by mirh** [comment]
Take note that until some refactoring in [early](https://lkml.org/lkml/2019/12/14/126) [2020](https://github.com/systemd/systemd/issues/17056), some "legit effort by bootloader" was also needed.

**Comment by maharmstone** [comment]
systemd contains EFI code? Why am I not surprised...

You can write EFI runtime services drivers that stay resident while the OS is running - that'd be one way to implement a thunking layer. Or you could just fake it, depending on exactly which runtime services Windows actually uses.

**Comment by mirh** [comment]
Systemd merged gummiboot a long time ago, but it could have been as well an entirely different project. 
They don't really share anything AFAIK and it's also probably the leanest bootloader out there.

**Comment by Juli199696** [comment]
Im also very interested in this Topic. Got an Z3736 Board with 32Bit Bios and 32bit Windows.

Also the reason i know that why many of that boards are 32bit is that in that time Intel annouced the Baytrail Atom chips the 64bit uefi was not ready for release and a little buggy. So the developers of the boards where forced to use the 32bit version. I dont know if that is true but thats what i have read in the internet about that topic. So are there any news about it?

**Comment by Juli199696** [comment]
Its not about better performance. Its about to be able to even run some Programms that only supports 64bit. Like Nvidia GeForce now or most newer games. I can even run 64bit apps on my Raspberry pi 4 Windows on ARM so please. Its so sad q.q

**Comment by 1nikolas** [comment]
> 
> 
> Its not about better performance. Its about to be able to even run some Programms that only supports 64bit. Like Nvidia GeForce now or most newer games. I can even run 64bit apps on my Raspberry pi 4 Windows on ARM so please. Its so sad q.q

Yeah that's what I said (you can read the discussion here). And apparently there is actually a performance benefit (see [here](https://github.com/maharmstone/quibble/issues/2#issuecomment-619434191)) but I have no idea about how any of this works

**Comment by supertiger1234** [comment]
Windows 11 won't have a 32-bit version. Has anyone confirmed if this method works?

**Comment by LoreBadTime** [comment]
Same problem here,got a 32bit tablet,someone found a solution?

**Comment by 1nikolas** [comment]
@supertiger1234 @LoreBadTime you can still use Windows 10 (32bit) and get latest updates til 2025

**Comment by LoreBadTime** [comment]
The main problem isn't windows 11,but that now 32bit apps are all discontinued,my main goal was to run 64 bit apps, fortunately i can run puppyLinux 64 with wine,but still a lot of apps gives problems even with wine

**Comment by 1nikolas** [comment]
> 
> 
> The main problem isn't windows 11,but that now 32bit apps are all discontinued,my main goal was to run 64 bit apps, fortunately i can run puppyLinux 64 with wine,but still a lot of apps gives problems even with wine

Try archlinux with this wine https://github.com/Frogging-Family/wine-tkg-git

**Comment by LoreBadTime** [comment]
The problem could be installing it, puppyLinux run entirely on RAM,and i can save data on external storage,so i don't need to install anything on the computer,if i try to fully install something i think that the bootloader will just not boot the os

**Comment by 1nikolas** [comment]
> 
> 
> The problem could be installing it, puppyLinux run entirely on RAM,and i can save data on external storage,so i don't need to install anything on the computer,if i try to fully install something i think that the bootloader will just not boot the os on the hard drive(happened to other users of the tablet)

Yeah installing 64bit Windows would be ideal but it seems that @maharmstone doesn't have the time to implement that. The issue is open for 1.5 years..

**Comment by supertiger1234** [comment]
@1nikolas I know that i can still use windows 10, but I just wanted to see windows 11 working on my old tablet ://

**Comment by 1nikolas** [comment]
> @1nikolas I know that i can still use windows 10, but I just wanted to see windows 11 working on my old tablet ://

I mean.. Running Windows 11 on these devices? Windows 11 requres 4gb ram and I don't think there is any 32bit locked device with this much ram. And remember: the reason the lock the bios to 32bit only is because of the ram (according to intel). So if this ever gets implumented your best bet would be a slimed down Windows 10 or Windows 7 (64bit)

**Comment by LoreBadTime** [comment]
My tabled has 4gb,and because of this limitation i can't even use all the ram,what a waste

**Comment by 1nikolas** [comment]
> My tabled has 4gb,and because of this limitation i can't even use all the ram,what a waste

Yeah that's dumb. In your case it will be useful. But in most cases these devices only have 1 or 2gb of ram (like mine)

**Comment by supertiger1234** [comment]
@1nikolas Windows 11 works even with 1gb ram on a VM.

**Comment by 1nikolas** [comment]
> 
> 
> @1nikolas Windows 11 works even with 1gb ram on a VM.

the fact that it runs doesn't mean it's usable. With no apps running it should be at like 80% ram. Anyway we can try that if this gets done some day

**Comment by supertiger1234** [comment]
@1nikolas mhm I don't really mind about how usable it will be to be honest 😂 I just wanna see it

**Comment by ShadowEO** [comment]
> > The problem could be installing it, puppyLinux run entirely on RAM,and i can save data on external storage,so i don't need to install anything on the computer,if i try to fully install something i think that the bootloader will just not boot the os on the hard drive(happened to other users of the tablet)
> 
> Yeah installing 64bit Windows would be ideal but it seems that @maharmstone doesn't have the time to implement that. The issue is open for 1.5 years..

It's more that it's a huge ask to implement. The Linux Kernel can do this because it takes over Runtime Services, but there's no such mechanism in Windows and the thunking between the binary architectures would need to be implemented outside of Windows. 

One way I see around this problem would be to use a bare-metal 64-bit client hypervisor (such as Xen or XenClient, _if_ one can find a good copy or get OpenXT to work on their hardware) and use that, as many of these machines do natively support VT-x and can run 64-bit VMs through that support. That does strip the user of GPU and certain hardware support, but you would essentially be running 64-bit Windows on the ia32-locked platform.

**Comment by supertiger1234** [comment]
@ShadowEO But can I interact with the os on the same device?

**Comment by ShadowEO** [comment]
> @ShadowEO But can I interact with the os on the same device?

I am unsure, I haven't been able to install OpenXT in a VM with nested virtualization to try it out. I don't want to install it on my main machine to try it because I need the machine for work.

XenClient was a bare-metal hypervisor that would indeed let you interact with the OS on the same device. OpenXT is supposed to be an open source replacement for XenClient, so I would assume.

**Comment by IminatoPo** [comment]
> > I have an older onda v919 z3975 atom that is similar. It is still a very capable machine with 4 core. would be great to run 64 bit windows or linux or even chromeos. But UEFI is crippled to 32 bit.
> 
> Running 64bit Windows would depend on this project.
> 
> Linux you can run 64bit (in fact you can't run 32bit linux on uefi afaik). On most of the distros you will need to add an efi file to the installer and then boot again to the live cd after the install to add that file to your installation in order for it to boot (you dont need to do this on fedora official and arch custom builds).
> 
> Chromeos you can run 32bit only but in my case, eventhough the laptop I have has a chromeos version, it doesn't boot no matter what I do..
> 
> Btw I have a Lenovo 100s-11iby

I have the x205ta, and it has the same chip with Lenovo 100s-11iby, I installed Manjaro KDE 64bit version followed this guide below:

**Comment by ShadowEO** [comment]
This feature would literally save running modern versions of Windows on these devices now. Microsoft seems to have stopped providing 32-bit versions of Windows as of the upcoming 11, meaning these arbitrarily bit-locked tablets are essentially either Windows 10 only, or Linux machines from this point on.

**Comment by 1nikolas** [comment]
> This feature would literally save running modern versions of Windows on these devices now. Microsoft seems to have stopped providing 32-bit versions of Windows as of the upcoming 11, meaning these arbitrarily bit-locked tablets are essentially either Windows 10 only, or Linux machines from this point on.

I mean, would these cheap devices be compatible with Windows 11 anyways? (Cpu gen, tpm, ram, etc) Unless you do a 'hacky' install which would probably be insecure.

I personally installed Linux on my device because it was running slow on Windows 10.

**Comment by supertiger1234** [comment]
Linux doesn't run well on my tablet. I get audio issues such as when plugging in the headphone jack, audio is outputted from the device and the headphones. I have no way to fix it either :(

**Comment by 1nikolas** [comment]
> Linux doesn't run well on my tablet. I get audio issues such as when plugging in the headphone jack, audio is outputted from the device and the headphones at the same time. I have no way to fix it either :(

Which distro did you try? My ideapad 100s sound works just fine. Maybe you need something with latest kernel (like Arch). Unless it's a hardware problem

**Comment by supertiger1234** [comment]
I Think I tried mint Linux or something. I also tried Android x86 with the same problem. Definitely not hardware since it works fine in windows 10

**Comment by 1nikolas** [comment]
Try using ArchLinux (or RebornOS for a more user friendly ui) and see if it works there. Maybe it was a bug fixed on the latest kernel (which mint and ubuntu do not use)

**Comment by supertiger1234** [comment]
tried installing rebornos os, it doesn't have bootai32.efi included so I downloaded it and put it in the correct file, it booted to the grub command line thing. I have tried to chainload or whatever but I get "unknown error". this seems complicated to install :/

**Comment by ShadowEO** [comment]
> > This feature would literally save running modern versions of Windows on these devices now. Microsoft seems to have stopped providing 32-bit versions of Windows as of the upcoming 11, meaning these arbitrarily bit-locked tablets are essentially either Windows 10 only, or Linux machines from this point on.
> 
> I mean, would these cheap devices even be compatible with Windows 11 anyways? (Cpu gen, tpm, ram, etc) Unless you do a 'hacky' install which would probably be insecure.
> 
> I personally installed Linux on my device because it was running slow on Windows 10. Let alone 11

I would like to point out that just because it's not considered "compatible" does not make it any less secure as long as there is indeed a TPM module, and yes, many of those devices do indeed have a TPM installed and available, and the processors are usually generally very capable, though the OEM decided to lock the system to running 32-bit only code.

@supertiger1234 This is perhaps not a good place to get into the topic of installing a linux distribution on one of these devices, look up "Intel Bay/Cherry Trail Atom Linux installation" ([here's one](https://gist.github.com/jfstenuit/09feac5ab0bff500db81ac9a56a48773) though)

The message you're receiving is either due to your PC being EXACTLY what I described in this topic earlier (IA32 UEFI on a 64-bit processor, which locks out bootx64.efi booting) or you aren't a 64-bit machine at all. If you know your processor IS 64-bit capable, then you're using the wrong version of Grub (you need the IA32-AMD64 build, not the AMD64 build itself)

**Comment by ShadowEO** [comment]
Your distribution probably instance the wrong grub package. For these machines, on Debian derived distros, the correct package is grub-efi-amd64:i386

The :i386 at the end is crucial, as this installs a 64-bit capable grub EFI binary that can run under 32-bit

**Comment by ShadowEO** [comment]
> @ShadowEO like I said, I already googled and tried what I you told me to do below. I just give up with this tablet. I'm gonna leave it on windows 10.
> 
> > tried installing rebornos os, it doesn't have bootai32.efi included so I downloaded it and put it in the correct file, it booted to the grub command line thing. I have tried to chainload or whatever but I get "unknown error". this seems complicated to install :/

It definitely is difficult, especially for installing a distribution. You could attempt Debian or Ubuntu, I believe they have this fix in the 64-bit version by default. As for creating a bootloader that could escalate us up to AMD64 mode despite the built-in UEFI firmware DOES appear possible, as the Windows-On-ARM team created a bootloader that escalates them from ARM32 to AARCH64.
